File: hIndexII/hIndexII.cpp
// clang++ hIndexII.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
  int hIndex(vector<int>& citations) {
    int left = 0, right = citations.size() - 1;
    while (left <= right) {
      int mid = left + (right - left) / 2;
      if (citations[mid] >= citations.size() - mid) {
	right = mid - 1;
      } else {
	left = mid + 1;
      }
    }
    return citations.size() - left;
  }
};

int main()
{
  Solution s;
  vector<int> vec = {0,1,3,5,6};
  cout << s.hIndex(vec);
  cout << endl;
  return 0;
}

File: 275.hIndexII/hIndexII.java
class Solution {
    public int hIndex(int[] citations) {
        int len = citations.length;
        if (len == 1) {
            if (citations[0] >= 1) return 1;
            else return 0;
        }
        int left = 0, right = len - 1;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (citations[mid] >= len - mid) {
                right = mid;
            } else {
                left = mid;
            }
        }
        if (citations[left] >= len - left) return len - left;
        else if (citations[right] >= len - right) return len - right;
        else return 0;
     }
}
